### ISSUE TITLE
Events hook

### ISSUE THREAD

**Issue by agracey** [epinio/epinio#593]
I would like a way for external components to register or stream events that happen in Epinio. 

These events would include:
- app lifecycle (maybe including healthcheck failures?)
- service lifecycle
- user login
- build pack progress

App logs, metrics, and tracing are delivered through their respective channels so don't need to be included.

**Comment by agracey** [comment]
Could this be done in an external component since most everything is done via k8s objects anyways?

I think we would just need the user that affected a change to be in the annotations or labels.

**Comment by agracey** [comment]
@jimmykarily Sounds like this will likely be needed for the UI work

**Comment by jimmykarily** [comment]
Yes but I think we can start with polling and improve from there. Unless we implement some "watchers" in the backend, we would simply be moving the polling from the front end (UI) to the backend. E.g. if we want to stream events about applications through websockets, something needs to gather the relevant events from Kubernetes. Since there is nothing monitoring Kubernetes for such changes, our backend code would have to poll the state of the various applications and present the results as a stream.

Since UI can start with polling and given the complexity of implementing such a solution in the backend, I would say we should start building a UI with what we have and start planning for this improvement for a future.

**Comment by jimmykarily** [comment]
I think the UI will be interested in events as they happen in the Epinio context, not Kubernetes. I mean, the UI is interested in the event of the app being staged. In the Kubernetes context this translates to a number of events (tekton pods spinning up, a PipelineRun is created etc).

It makes more sense to implement this mechanism in Epinio. I have some ideas on how this could be implemented in a simple way. But let's check we agree on the high level first. @richard-cox I'm thinking something like:

- UI subscribes to topics (each page may be interested in different things) by calling a websocket endpoint with some params (e.g. the topics)
- Epinio sends events there on the various topics
- Events are kept small. Only the information needed for the UI to make the appropriate call to gather more information. 

E.g.
- UI receives an event that a service has been updated
- UI makes a GET request to get the service and update it in the view if needed.

We can discuss the details, but is this what you have in mind? I know there is another event system you are consuming in Rancher so maybe you already have a preference on how this should work.

**Comment by jimmykarily** [comment]
A thing to keep in mind is that simple solutions that are contained within the Epinio binary will probably break HA-bility of the Epinio server. Up to now, the single source of truth was Kubernetes and that's where each and every instance of the Epinio looked (provided you have more than one instances).

Events that happen on another instance of Epinio should somehow make it to subcribers listening on other instances. This requires a centralized pub/sub component (or a very complicated syncing mechanism). Such a component must also be able to scale which means we probably don't want to implement that ourselves. There are off the shelf pub/sub solutions but they will be yet another components to maintain. In any case, the event system could be made optional (if it's there we use it otherwise ignore it).

**Comment by agracey** [comment]
I'm cool with this being in scope for one of the upcoming milestones if there's room. 

I wonder if it might be better to have a websocket server that's in a separate pod that each of the epinio servers push events to. This new component wouldn't have to be HA.

**Comment by agracey** [comment]
The websocket server would likely need to dedup some event types so it's likely not trivial

**Comment by richard-cox** [comment]
Dashboard / Rancher API works by...
- Dashboard opens socket(s) with Rancher for push events
- Dashboard requests kube resources of various types
- Rancher pushes updates to these resources that have previously been requested to the Dashboard via socket
  - I think these come in two forms. Single resources (someone has edited it) or all resources of that type (create, delete) 

Would there be the two options then?
- Similar to the above, either UI registers interest or interest is taken from API requests and any changes to these resources are pushed back to interested clients (via socket)
- A generic log of 'app A deployed, service B modified, namespace x deleted' is provided by the backend... which the frontend then interprets and makes the appropriate update requests (fetch a specific resource, fetch all resources of that type). Frontend would need to carefully manage throttling/debouncing for individual clients.

The second would be nicer (there's value in showing that log to users), however the backend would get DDOS if there were a lot of clients.

**Comment by jimmykarily** [comment]
> This new component wouldn't have to be HA.

At some point all the components should be HA. While we haven't done any specific work for that or any testing, we didn't intentionally break HA-bility either. I'm ok to start with an implementation that is not HA as long as we have a plan how to make it HA. For example, if we decided to use Redis as the pub/sub backend, we could start with a simple non-HA deployment. If a user want's HA Redis, there are ways to setup one and it's not Epinio's responsibility. That's the same with all the "external" components for Epinio.

**Comment by jimmykarily** [comment]
We had a call with @richard-cox @catherineluse and @agracey about this.

This is the result of our conversation:

There should be a single point of truth for events. That source of truth should be HA-able. This means we don't want to write it ourselves because of the complexity in syncing between the various instances. There are third party solutions that we could use but there are options:

1. Put that "component X" behind the Epinio instances
  Epinio takes care of authentication like usual. Consumers subscribe to events by talking to the Epinio API.
2. Put "component X" next to Epinio
  Consumers subscribe to events directly on "component X". Epinio instances send events to "component X". The problem is, how do consumers authenticate against "component X"? Ideally Epinio should still be responsible for authentication

Option 2 is better because it requires less code in Epinio. All instances of Epinio simply send events to "component X" if one is there. The authentication problem may be solvable by adding another component in front of "component X" (or it could even be just another endpoint on Epinio API). That component/endpoint would act as a reverse proxy which would first do authentication and then simply forward the request to "component X".

We summarized this idea in this image:

![epinio events](https://user-images.githubusercontent.com/2794419/149182893-79872720-699b-4836-bf22-c1d94daacc0c.jpg)

This is raw brain storming so it deserves some though and probably a spike.

**Comment by richard-cox** [comment]
@catherineluse  and I have put together a very rough spec of requirements from a UI perspective

Support UI Features
1. A generic event stream view where the user can see events in real time
2. UI to ensure local cache of resources is kept up to date, thus removing resource polling

Event types should cover use cases such as...
- useful for a developer 
- useful for a cluster maintainer
- cover the means to manage the local ui resource cache

## Types

## Must Have
- Changes to all resource types
    - CREATE
    - EDIT
    - DELETE

## Nice to have 
- User log in

## Suggestions
- Application restarted/restaged
- Service bound


## Streaming Features

### Must have
- Start/Stop (Pause debatable)
- Events should contain
  - timestamp
  - full version of resource/s that have changed (if event relates to a resource type)
- Stream should only return events for types that have been requested. Two examples...
  - Connection 1 registers events related to applications and namespaces resource types. Events related to services are not sent
  - Events streaming view disables certain spammy event types or only enables a single type

### Nice to have
- Filter events not just by type but also namespace (this can be done in the frontent but would be nicer by backend)

**Comment by agracey** [comment]
@enrichman This seems (to me) like it would be touching the same code paths as the access control work. Would it be possible to take a first pass at this in the same chunk of work?

**Comment by agracey** [comment]
Open question: are there any components that Rancher has to make this streamlined?

**Comment by agracey** [comment]
A first pass could just be:
- User tried to do something (allowed or not)
- Error occurred
- Build triggered
